pkg/controller/volume/attachdetach/metrics/metrics.go

```diff
@@ -0,0 +1,157 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+var (
+	inUsePVCMetricDesc = prometheus.NewDesc(
+		prometheus.BuildFQName("", "storage_counter", "pvc_in_use"),
+		"Measure number of pvcs in use",
+		[]string{"node", "volume_plugin"}, nil)
+)
+var registerMetrics sync.Once
+
+type pvcInUseCollector struct {
+	pvcLister       corelisters.PersistentVolumeClaimLister
+	podLister       corelisters.PodLister
+	pvLister        corelisters.PersistentVolumeLister
+	volumePluginMgr *volume.VolumePluginMgr
+}
+
+type nodePVCCount map[types.NodeName]map[string]int
+
+// Register registers pvc's in-use metrics
+func Register(pvcLister corelisters.PersistentVolumeClaimLister,
+	pvLister corelisters.PersistentVolumeLister,
+	podLister corelisters.PodLister,
+	pluginMgr *volume.VolumePluginMgr) {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(newPVCInUseCollector(pvcLister, podLister, pvLister, pluginMgr))
+	})
+
+}
+
+func (pvcInUse nodePVCCount) add(nodeName types.NodeName, pluginName string) {
+	nodeCount, ok := pvcInUse[nodeName]
+	if !ok {
+		nodeCount = map[string]int{}
+	}
+	nodeCount[pluginName]++
+	pvcInUse[nodeName] = nodeCount
+}
+
+func newPVCInUseCollector(
+	pvcLister corelisters.PersistentVolumeClaimLister,
+	podLister corelisters.PodLister,
+	pvLister corelisters.PersistentVolumeLister,
+	pluginMgr *volume.VolumePluginMgr) *pvcInUseCollector {
+	return &pvcInUseCollector{pvcLister, podLister, pvLister, pluginMgr}
+}
+
+func (collector *pvcInUseCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- inUsePVCMetricDesc
+}
+
+func (collector *pvcInUseCollector) Collect(ch chan<- prometheus.Metric) {
+	nodePVCMap := collector.getPVCUseByNode()
+	for nodeName, pluginCount := range nodePVCMap {
+		for pluginName, count := range pluginCount {
+			metric, err := prometheus.NewConstMetric(inUsePVCMetricDesc,
+				prometheus.GaugeValue,
+				float64(count),
+				string(nodeName),
+				pluginName)
+			if err != nil {
+				glog.Warningf("Failed to create metric : %v", err)
+			}
+			ch <- metric
+		}
+	}
+}
+
+func (collector *pvcInUseCollector) getPVCUseByNode() nodePVCCount {
+	pods, err := collector.podLister.List(labels.Everything())
+	if err != nil {
+		glog.Errorf("Error getting pod list")
+		return nil
+	}
+
+	nodePVCMap := make(nodePVCCount)
+	for _, pod := range pods {
+		if len(pod.Spec.Volumes) <= 0 {
+			continue
+		}
+
+		nodeName := types.NodeName(pod.Spec.NodeName)
+		if nodeName == "" {
+			continue
+		}
+		for _, podVolume := range pod.Spec.Volumes {
+			volumeSpec, err := collector.CreateVolumeSpec(podVolume, pod.Namespace)
+			if err != nil {
+				continue
+			}
+			volumePlugin, err := collector.volumePluginMgr.FindPluginBySpec(volumeSpec)
+			if err != nil {
+				glog.V(3).Infof("Error finding volume plugin for : %v", volumeSpec)
+				continue
+			}
+			nodePVCMap.add(nodeName, volumePlugin.GetPluginName())
```

Comment:
Is this to help with the attachable limit count? If so, do we want to use the attachable limit resource name instead of the plugin name?  For CSI, I think this just returns "kubernetes.io/csi"

Reply:
This is not just for attachable limit count. It is useful for capacity planning in multi-tenant clusters and perhaps there are other usage too. In a multi-tenant cluster one typically has many more PV/PVCs than actually in-use PVCs. This metric obviously helps in determining if a cluster admin needs to spin up new nodes to accomodate in-use PVCs that takes into account both attachable limit or limits that apply in general to volume types that we don't even consider "attachable" right now.

I agree with unfortunate situation with CSI plugin name and `GetPluginName` is not a satisfactory solution. But all storage metrics are affected by this problem, because they all use `GetPluginName` as a label. I have filed https://github.com/kubernetes/kubernetes/issues/64590 to solve this. I think we are going to need additional function call or something for CSI plugins.